### PR TITLE
Add loaded blocks to transforming_liquid-queue similar to updateLiquid

### PR DESCRIPTION
**Pull request by MillersMan**
This adds liquid nodes of blocks loaded from data-base to ServerMap::m_transforming_liquid with the same abstract logic as in MapGen::updateLiquid and has the following fixes:
- Liquid from above will start flowing into a newly loaded block
- Unfinished transformations inside the block will resume after loading
- Liquid in newly loaded blocks will start flowing to next lower block if that was already loaded
- Only nodes that actually need to be transformed are added to the queue (unlike MapGen::updateLiquid which always adds the whole topmost plane)

### CODE REVIEW

**Comment by MillersMan** [comment]
I've tested with https://github.com/minetest/minetest/files/101456/Just.Test.zip
It looks good so far. However in very few cases some nodes aren't transformed. There is nothing special about those nodes, as they aren't at block borders and after a server restart those are transformed like all others.

**Comment by Zeno-** [comment]
Looks extremely promising

**Comment by kwolekr** [comment]
We already discussed this on IRC, the approach here is not a good one and has been done in multiple PRs prior to this one.

**Comment by paramat** [comment]
Related #2977 
See coments form here onwards https://github.com/minetest/minetest/pull/2977#issuecomment-171900572 the concern is too many additions to the liquid queue. Maybe we should consider just updating the nodes in the lowest layer of nodes in a mapblock, as these are the ones most likely to get stuck?

**Comment by kwolekr** [comment]
Could you profile to see how many nodes get added to the liquid transform queue on average per block load?

**Comment by MillersMan** [comment]
I did tests with valley mapgen and the results so far are:

- Worst case: Shallow ocean ~100 transformations per block (would expect ~500 for the blocks themselves, but average is over all blocks including air, ground and 100% water which all have no transformations)
- Surface and ground of deep oceans: below 100 per block (don't have exact values at it is spoiled with result from worst case scenario, would expect ~250 but as before the outcome is average)
- Valleys with rivers: Below 10 transformations per block (would expect 2 per column if column is higher than one node, otherwise it would be one. So here it is the same, blocks without liquids drop the rate)
- Deserts or other areas without liquids: No transformations at all

**Comment by MillersMan** [comment]
It would even be possible to reduce it close to zero if surrounding nodes are checked, too. But this might bring a small burden during the scan phase (will do a separate performance tests for this) would expect about 5% extra scan cost at loading time to only queue nodes for transformation that will change the environment,

**Comment by Fixer-007** [comment]
I've tried it with #4065 and have not seen any issues with reflow yet, it activates nicely, even huge amounts, rarely you need to revisit place (fly away and come back) to fix rare pieces.
I've looked both "just test" glitches and mgvalleys.

**Comment by MillersMan** [comment]
The issue observed by Fixer with the gaps will be solved when the #4065 changes are also applied to the loading algorithm. The problem is in fact that the upper node wasn't loaded yet so there was a ignored-node above that was interpreted like air. 4065 solves this by not activating blocks below ignored ones at all

File: src/emerge.cpp [hunk 1]
```diff
@@ -595,6 +596,15 @@ EmergeAction EmergeThread::getBlockOrStartGen(
 }
 
 
+void EmergeThread::finishLoad(MapBlock* block)
+{
+	MutexAutoLock envlock(m_server->m_env_mutex);
+	
+	if(block)
```
Comment (nerzhul) [inline src/emerge.cpp]: space, and test this before locking to reduce the useless locking (which costs).

**Comment by Fixer-007** [comment]
@MillersMan Any progress on this PR?

**Comment by paramat** [comment]
@MillersMan are you able to finish this for release? (feature freeze early December).

**Comment by MillersMan** [comment]
@paramat There are a few bigger issues that have to be addressed. Loaded blocks are far more difficult than initially created blocks. Those are the issues on my list:
* ignored nodes above or on the side should prevent flowing away as they might support the stream when loaded (affects normal flow algorithm)
* ignored nodes below should not start spreading (affects normal flow algorithm)
* after loading: air below liquids should start flow
* after loading: liquid above solid or source nodes should spread horizontally to air or to liquid nodes with low level
* reflow has to respect the border-nodes of blocks next to the loaded block
* reflow has to be applied to all loaded blocks, currently some blocks are loaded via different function calls that don't result in a reflow-scan

I think all of them are solvable but my original approach would make this to complex and slow. Maybe i should scan in three separate passes, one for each axis. This might result in a node being added three times to the liquid-update-queue in the very unlikely worst case scenario but it should have an acceptable overall performance.

**Comment by paramat** [comment]
> ignored nodes below liquid nodes should not result in horizontal spreading (affects normal flow algorithm), i think this was already implemented

Yes i did that.

If you want a simplified solution the main problem is reflowing liquids in the lowest layer of nodes in a block. This would enable water columns to drop a long distance and be used for travel. I would be happy with just that fixed.

**Comment by Fixer-007** [comment]
Quoting:
>#4061 is updated and in the final state (or at least close to)
Testing time!

File: src/map.cpp [hunk 2]
```diff
@@ -1336,10 +1338,14 @@ void Map::transformLiquids(std::map<v3s16, MapBlock*> &modified_blocks)
 							flowing_down = true;
 					} else {
 						neutrals[num_neutrals++] = nb;
-						// If neutral below is ignore prevent water spreading outwards
-						if (nb.t == NEIGHBOR_LOWER &&
-								nb.n.getContent() == CONTENT_IGNORE)
-							flowing_down = true;
+						if(nb.n.getContent() == CONTENT_IGNORE) {
+							// If neutral below is ignore prevent water spreading outwards
+							// otherwise prevent flowing away as neutral node might be a source
```
Comment (paramat) [inline src/map.cpp]: If not already please split comment at 80 columns (with tab size 4).

File: src/map.cpp [hunk 3]
```diff
@@ -1336,10 +1338,14 @@ void Map::transformLiquids(std::map<v3s16, MapBlock*> &modified_blocks)
 							flowing_down = true;
 					} else {
 						neutrals[num_neutrals++] = nb;
-						// If neutral below is ignore prevent water spreading outwards
-						if (nb.t == NEIGHBOR_LOWER &&
-								nb.n.getContent() == CONTENT_IGNORE)
-							flowing_down = true;
+						if(nb.n.getContent() == CONTENT_IGNORE) {
```
Comment (paramat) [inline src/map.cpp]: Space after 'if'

File: src/map.cpp [hunk 4]
```diff
@@ -1336,10 +1338,14 @@ void Map::transformLiquids(std::map<v3s16, MapBlock*> &modified_blocks)
...
+							// If neutral below is ignore prevent water spreading outwards
+							// otherwise prevent flowing away as neutral node might be a source
+							if(nb.t == NEIGHBOR_LOWER)
```
Comment (paramat) [inline src/map.cpp]: Space after 'if'

File: src/map.cpp [hunk 5]
```diff
@@ -2961,8 +2974,11 @@ void ServerMap::loadBlock(std::string *blob, v3s16 p3d, MapSector *sector, bool
 		block->deSerialize(is, version, true);
 
 		// If it's a new block, insert it to the map
-		if(created_new)
+		if(created_new) {
```
Comment (paramat) [inline src/map.cpp]: Space after 'if'

File: src/reflowscan.cpp [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+inline MapBlock *ReflowScan::lookupBlock(int x, int y, int z)
+{
+	int bx = (MAP_BLOCKSIZE + x) / MAP_BLOCKSIZE;
+	int by = (MAP_BLOCKSIZE + y) / MAP_BLOCKSIZE;
+	int bz = (MAP_BLOCKSIZE + z) / MAP_BLOCKSIZE;
+	int idx = (bx + (by * 9) + (bz * 3));
+	MapBlock *result = m_lookup[idx];
+	if(!result && (m_lookup_state_bitset & (1 << idx)) == 0) {
```
Comment (paramat) [inline src/reflowscan.cpp]: Space after 'if'

File: src/reflowscan.cpp [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+		m_lookup_state_bitset |= (1 << idx);
+	}
+	return result;
+}
+
+inline bool ReflowScan::isLiquidFlowableTo(int x, int y, int z)
+{
+	bool valid_position;
+	MapBlock *block = lookupBlock(x, y, z);
+	if(block) {
```
Comment (paramat) [inline src/reflowscan.cpp]: Space after 'if'

File: src/reflowscan.cpp [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+		isLiquidFlowableTo(x, y, z + 1);
+}
+
+void ReflowScan::scanColumn(int x, int z)
+{
+	bool valid_position;
+
+	// Is the column inside a loaded block?
+	MapBlock *block = lookupBlock(x, 0, z);
+	if(!block)
```
Comment (paramat) [inline src/reflowscan.cpp]: Space after 'if'

File: src/reflowscan.cpp [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+	// Scan neighbouring columns
+	for (s16 i = 0; i < MAP_BLOCKSIZE; i++) {
+		scanColumn(i, -1);
+		scanColumn(i, MAP_BLOCKSIZE);
+		scanColumn(-1, i);
+		scanColumn(MAP_BLOCKSIZE, i);
+	}
+}
+
+inline MapBlock *ReflowScan::lookupBlock(int x, int y, int z)
```
Comment (paramat) [inline src/reflowscan.cpp]: Is this simple enough to be inline? Likewise for functions below.

File: src/reflowscan.cpp [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+	bool valid_position;
+	MapBlock *block = lookupBlock(x, y, z);
+	if(block) {
+		int dx = (MAP_BLOCKSIZE + x) % MAP_BLOCKSIZE;
+		int dy = (MAP_BLOCKSIZE + y) % MAP_BLOCKSIZE;
+		int dz = (MAP_BLOCKSIZE + z) % MAP_BLOCKSIZE;
+		MapNode node = block->getNodeNoCheck(dx, dy, dz, &valid_position);
+		if (node.getContent() != CONTENT_IGNORE) {
+			const ContentFeatures &f = m_ndef->get(node);
+			if (f.floodable && f.liquid_type != LIQUID_SOURCE)
```
Comment (paramat) [inline src/reflowscan.cpp]: Is the check for liquid source needed? i doubt liquids are 'floodable'.

File: src/reflowscan.cpp [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+	// Is the column inside a loaded block?
+	MapBlock *block = lookupBlock(x, 0, z);
+	if(!block)
+		return;
+
+	MapBlock *above = lookupBlock(x, MAP_BLOCKSIZE, z);
+	int dx = (MAP_BLOCKSIZE + x) % MAP_BLOCKSIZE;
+	int dz = (MAP_BLOCKSIZE + z) % MAP_BLOCKSIZE;
+
+	bool was_ignored, was_liquid;
```
Comment (paramat) [inline src/reflowscan.cpp]: 'was_ignored' has a different meaning to 'was_ignore', do you mean the node has been ignored (not processed or bothered with), or that the node was 'content ignore'?

File: src/reflowscan.cpp (hunk 9 above)
Comment (MillersMan) [inline src/reflowscan.cpp]: The inline keyword is just a hint to the compiler that it doesn't have to follow. I've added it because inlining might allow optimizations like merging division and modulo (ok, they likely already got replaced with shifts, as MAP_BLOCKSIZE is a power of two) or eliminating some operations like adding MAP_BLOCKSIZE to x/y/z twice.

In short: IMHO having inline there might bring more advantages than removing them but they aren't absolutely necessary

File: src/reflowscan.cpp [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+		was_ignored = true;
+		was_liquid = false;
+	}
+	bool was_checked = false;
+	bool was_pushed = false;
+
+	for (s16 y = MAP_BLOCKSIZE - 1; y >= 0; y--) {
+		MapNode node = block->getNodeNoCheck(dx, y, dz, &valid_position);
+		const ContentFeatures &f = m_ndef->get(node);
+		bool is_ignored = node.getContent() == CONTENT_IGNORE;
```
Comment (paramat) [inline src/reflowscan.cpp]: Perhaps 'is_ignore' instead?

File: src/reflowscan.cpp (hunk 10 above)
Comment (MillersMan) [inline src/reflowscan.cpp]: I've seen the expected effects so I think liquids are floodable. But as it might be an error in this PR if they weren't so I'll go and double-check it.

File: src/reflowscan.cpp (hunk 12 above)
Comment (MillersMan) [inline src/reflowscan.cpp]: Maybe, but if it is set I'll skip the rest of the computation and just store it as was_ignored for the next cycle. So it isn't wrong either

File: src/map.cpp (hunk 3 above)
Comment (MillersMan) [inline src/map.cpp]: Ok, I'll fix those

File: src/reflowscan.cpp [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+
+		was_liquid = is_liquid;
+		was_ignored = is_ignored;
+	}
+
+	MapBlock *below = lookupBlock(x, -1, z);
+	if(below) {
+		MapNode node = below->getNodeNoCheck(dx, MAP_BLOCKSIZE - 1, dz, &valid_position);
+		const ContentFeatures &f = m_ndef->get(node);
+		bool is_ignored = node.getContent() == CONTENT_IGNORE;
```
Comment (paramat) [inline src/reflowscan.cpp]: 'is_ignore'?

File: src/reflowscan.cpp (hunk 11 above)
Comment (MillersMan) [inline src/reflowscan.cpp]: Both, I don't do any computation for CONTENT_IGNORE nodes as adding them to the queue would not have any effect thanks to the anti-gap code in transformLiquids. But maybe was_ignore fits better... would then be the same for is_ignore

Comment (paramat) [inline src/reflowscan.cpp]: Ok no problem then, no need to change it.

File: src/reflowscan.cpp (hunk 9 above)
Comment (paramat) [inline src/reflowscan.cpp]: That's fine, you know more about inlining thanme.

**Comment by Fixer-007** [comment]
Big :+1: from me, I've checked with special maps and on fresh world and was very satisfied.

File: src/reflowscan.cpp (hunk 10 above)
Comment (paramat) [inline src/reflowscan.cpp]: I added 'floodable' quite recently to mean an airlike node that liquids can flow into and replace.
'floodable = true' is not in liquid node defs.

**Comment by paramat** [comment]
As far as i can tell the code looks good, but i'm not sure i understand it deeply enough to officially approve.

File: src/reflowscan.cpp (hunk 10 above)
Comment (paramat) [inline src/reflowscan.cpp]: I think the default is 'floodable = false'.

Comment (paramat) [inline src/reflowscan.cpp]: Indeed it is.

Comment (paramat) [inline src/reflowscan.cpp]: So liquid nodes are 'floodable = false'

File: src/reflowscan.cpp [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+	if (below) {
+		MapNode node = below->getNodeNoCheck(dx, MAP_BLOCKSIZE - 1, dz, &valid_position);
+		const ContentFeatures &f = m_ndef->get(node);
+		bool is_ignore = node.getContent() == CONTENT_IGNORE;
+		bool is_liquid = f.isLiquid();
+
+		if (is_ignore || was_ignore || is_liquid == was_liquid) {
+			// Neither topmost node of liquid column nor topmost node below column
+		} else if (is_liquid) {
+			// This is the topmost node in the column and might wan't to flow away
```
Comment (paramat) [inline src/reflowscan.cpp]: wan't -> want

**Comment by paramat** [comment]
Ok i mostly understand this and it looks good, there maty be some more technical aspects i don't understand so i'm giving half a +1.

**Comment by Fixer-007** [comment]
remove WIP?

File: src/reflowscan.cpp [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+	bool valid_position;
+	MapBlock *block = lookupBlock(x, y, z);
+	if (block) {
+		int dx = (MAP_BLOCKSIZE + x) % MAP_BLOCKSIZE;
+		int dy = (MAP_BLOCKSIZE + y) % MAP_BLOCKSIZE;
+		int dz = (MAP_BLOCKSIZE + z) % MAP_BLOCKSIZE;
+		MapNode node = block->getNodeNoCheck(dx, dy, dz, &valid_position);
+		if (node.getContent() != CONTENT_IGNORE) {
+			const ContentFeatures &f = m_ndef->get(node);
+			if (f.floodable && f.liquid_type == LIQUID_FLOWING)
```
Comment (paramat) [inline src/reflowscan.cpp]: Should be 'fllodable' OR liquid flowing

**Comment by paramat** [comment]
screenshot_20161129_015555

Tested that stuck water columns under floatlnds reflow with this PR.
Since i've tested this, mostly understand the code and since Fixer has tested thoroughly i feel i can give an official :+1: But another dev should check the code.

**Comment by Zeno-** [comment]
:+1: if @MillersMan can clarify that the issues described in the comment on 26 Oct have been addressed

**Comment by MillersMan** [comment]
@Zeno- Everything described in my comment on Oct 25 is fixed. This includes the water-columns mentioned in the paramat's comment on Oct 26.

**Comment by paramat** [comment]
Great.

**Comment by kwolekr** [comment]
Wow, this looks fantastic.  I'm sorry but I haven't had the time to actually do a review on this, since it does actually require some time and effort to carry out.

Would just like to point out a few minor style errors to fix on merge.

File: src/reflowscan.cpp [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+
+
+ReflowScan::ReflowScan(Map *map, INodeDefManager *ndef) :
+	m_map(map),
+	m_ndef(ndef),
+	m_liquid_queue(NULL)
+{
+}
+
+void ReflowScan::scan(MapBlock * block, UniqueQueue<v3s16> *liquid_queue)
```
Comment (kwolekr) [inline src/reflowscan.cpp]: Asterisk should be attached to block

File: src/reflowscan.h [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+#include "util/container.h"
+#include "irrlichttypes_bloated.h"
+
+class INodeDefManager;
+class Map;
+class MapBlock;
+
+class ReflowScan
+{
```
Comment (kwolekr) [inline src/reflowscan.h]: This should be on the previous line

File: src/reflowscan.h [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+public:
+	ReflowScan(Map *map, INodeDefManager *ndef);
+	void scan(MapBlock *block, UniqueQueue<v3s16> *liquid_queue);
+private:
+	MapBlock *lookupBlock(int x, int y, int z);
+	bool isLiquidFlowableTo(int x, int y, int z);
+	bool isLiquidHorizontallyFlowable(int x, int y, int z);
+	void scanColumn(int x, int z);
+private:
```
Comment (kwolekr) [inline src/reflowscan.h]: This should be a blank line

File: src/map.cpp [hunk 19]
```diff
@@ -174,7 +175,7 @@ bool Map::isNodeUnderground(v3s16 p)
 bool Map::isValidPosition(v3s16 p)
 {
 	v3s16 blockpos = getNodeBlockPos(p);
-	MapBlock *block = getBlockNoCreate(blockpos);
+	MapBlock *block = getBlockNoCreateNoEx(blockpos);
```
Comment (kwolekr) [inline src/map.cpp]: Huh, this isn't related to the PR... should it be a separate commit instead, despite its size?

**Comment by paramat** [comment]
@MillersMan splitting into 2 commits might be best as suggested.

File: src/map.cpp (hunk 19 above)
Comment (paramat) [inline src/map.cpp]: Looks like it is a seperate commit already.

**Comment by paramat** [comment]
@MillersMan if you fix up the style issues i'll merge this in 12 hours. If you have no time to update i'll alter on merge.

**Comment by MillersMan** [comment]
@paramat Sorry wanted to take a look on the issue earlier but there was another urgent issue. Would be grateful if you could do this on merge.

File: src/reflowscan.h (hunk 17 above)
Comment (MillersMan) [inline src/reflowscan.h]: Really? Typically class, struct and function bodys begin on a separate line while other blocks start on the same line as the declaration. Haven't seen any place in MT source where it is different.

**Comment by paramat** [comment]
No problem.

**Comment by paramat** [comment]
2829742ae8869276f000bdb64a9f14977e70038b
cc36f5e99ae46cf2fd370dde8311f139169a4c43
9714cdcf4ba51b44cca72e133ec5ef2a3c9f50f6

**Comment by Fixer-007** [comment]
Forgot to say. I waited for this for many years, I can't believe it is done and in. Good job 🥇
